### ISSUE TITLE
The fennel stdio client breaks Neovim when using some readline config

### ISSUE THREAD

**Issue by nicopap** [Olical/conjure#166]
First: thank you for making conjure.

Using the fennel stdio client, I found that neovim would get into a weird state that can only finish in a crash.

Good news, with a default setup, this doesn't happen.

### How to reproduce

* install lua, fennel, lua readline via luarocks. Setup environment so that lua finds the luarocks module
* add `let g:conjure#filetype#fennel = "conjure.client.fennel.stdio"` to your neovim config
* write the following snippet in your `~/.inputrc`:

```
$if fennel
    set editing-mode vi
    set show-mode-in-prompt on
    set vi-ins-mode-string "\1\e[3 q\2\1\015\2fennel"
    set vi-cmd-mode-string "\1\e[1 q\2\1\015\2fennel"
$endif
```

* make sure fennel found the readline lib by opening the repl with `fennel` and seeing there is no messages about readline anymore
* open a fennel file with neovim
* ahah, neovim bork, only way out is Ctrl-C

### My workaround

The error comes from my exotic readline config. Since I only setup readline with lua5.1 and not other versions, I can fix this by using a different lua version for fennel by adding the `--lua /path/to/luaVer` command line switch to the fennel command. What I did:

`let g:conjure#client#fennel#stdio#command = "fennel --lua /usr/bin/lua5.4"`

It's questionable whether you need the default fennel repl when you have conjure. But it's useful to report this bug so that when someone stumble upon this, they have a chance of finding the solution in the conjure bug tracker instead of having to waste an hour figuring things out :P

### Solution

I can't think of a "good" solution to this. I have many workarounds that come to mind, but they all involve unexpected side effects that probably leads to more problems than solutions:
* Remove the readline library from the `LUA_PATH` when invoking fennel from neovim
* Do not install/uninstall the readline lua library
* I have my suspicions that what is breaking neovim are the terminal escape sequences, so filtering them out in conjure might help. But that's not a good solution

The **real solution** to this is to add to the fennel executable a `--no-funny-stuff` cli switch to disable any user-specific funny stuff and keep a consistent interface so that conjure and other software can rely on it. But that involves making an upstream change to fennel, and I'm very lazy. On top of that, fennel is probably too young to commit to a "consistant CLI for automatic consumption".

Well, not a very actionable bug report, I'm sorry. But I just wanted to make sure people with the same issues are not stuck with no hopes of fixing it.

Again, my thanks and accolades for making conjure.

**Comment by nicopap** [comment]
Now that I think about it. Adding a line to `doc/conjure-client-fennel-stdio.txt` explaining the problem and providing workarounds would be a good temporary solution. I'll create a PR

**Comment by Olical** [comment]
So I think the comment in the doc file is a good start, thanks for that! I think a possible solution will be finding a way to disable readline for this REPL, possibly through an environment variable. I saw some people saying you can set your `TERM` to something that doesn't exist, like `TERM=conjure`, which readline can't support so it won't kick in.

**Comment by Olical** [comment]
I've been trying to get readilne working with my fennel CLI but i can't seem to do it :sweat_smile: maybe it's a NixOS thing.

I'll get it working at some point, I just want to try setting the TERM environment variable to a nonsense string to see if it disables readline stuff while keeping the program working. If so, I'll make some nonsense TERM the default with an option to disable that if you ever needed to.

**Comment by nicopap** [comment]
Your comment gave me a lead. So setting a bogus `TERM` doesn't work. But reading the readline man page, I found that you can overwrite the config using the `INPUTRC` env variables. By setting it to `INPUTRC=/dev/null`, it doesn't break neovim anymore.

Changing the `env` for spawned processes using `vim.loop.spawn` is actually poorly documented. But I found [this very helpfull page on libuv](https://nikhilm.github.io/uvbook/processes.html#spawning-child-processes), which tells us we simply need to provide an array of string in the form of `VAR=value`, so adding a `:env ["INPUTRC=/dev/null"]` to the table passed to `uv.spawn` in fnl/conjure/remote/stdio.fnl does the trick. I've tested it locally and it's promising.

**Comment by nicopap** [comment]
Currently reviewing the develop branch with the merged PR. I'm having troubles. Will update later.

**Comment by nicopap** [comment]
The patch doesn't work... I must have not tested properly.

**Comment by Olical** [comment]
What about this commit 8d5342aa7e8b63267f14e80a600bcdd00bd16ac5 which is the one directly after I merged your change, before I fiddled with it? So this is the version that starts with a completely blank environment, I added the changes to make it merge the `INPUTRC` override into your original environment.

Maybe it's my addition which isn't helping?

**Comment by nicopap** [comment]
Excellent suggestion. It seems 8d5342a indeed works. But I found that providing an empty `env` breaks a lot of things. Including loading the readline library. Since it also removes all the PATH, LUA_PATH etc. required to find the lua readline library.

**Comment by nicopap** [comment]
I'll experiment a bit more.

**Comment by Olical** [comment]
Ah interesting, well the code that overrides things is here https://github.com/Olical/conjure/blob/4604d440eea09f9edfa9d18ea1848b2fd8d6827f/fnl/conjure/remote/stdio.fnl#L111-L113 - maybe that's not working at all like I thought? Could be my fault with the additions I made! Tried to make it more ergonomic so we could extend / add more things where required, might not be working as expected.

Hopefully that point in the code is a good place for you to alter things.
